File: packages/persistent-storage-node/src/persistentStorage.ts
import { SegmentedLog } from '@krupton/rust-rocksdb-napi'; // ← your NAPI binding
import { mkdirSync } from 'node:fs';
import { dirname, join } from 'node:path';

export function normalizeIndexDir(subIndex: string): string {
  return subIndex
    .toLowerCase()
    .replace(/[^a-z0-9_]/g, '_')
    .replace(/_{2,}/g, '_')
    .replace(/^_+|_+$/g, '');
}

export type BaseStorageRecord = Record<string, unknown>;

export type StorageRecord<T extends BaseStorageRecord> = {
  timestamp?: number;
} & Omit<T, 'timestamp' | 'id'>;

export type StorageRecordReturn<T extends BaseStorageRecord> = {
  id: number;
  timestamp: number;
} & Omit<T, 'timestamp'>;

type WriteRecordParams<T extends BaseStorageRecord> = {
  record: StorageRecord<T>;
  subIndex: string;
};

type WriteRecordsParams<T extends BaseStorageRecord> = {
  records: StorageRecord<T>[];
  subIndex: string;
};

type ReplaceRecordParams<T extends BaseStorageRecord> = {
  record: StorageRecordReturn<T>;
  subIndex: string;
  id: number;
};
type ReplaceLastRecordParams<T extends BaseStorageRecord> = {
  record: StorageRecord<T>;
  subIndex: string;
};
type ReadFullPageParams = {
  subIndex: string;
};

type ReadRangeParams = {
  subIndex: string;
  fromId: number;
  count: number;
};
type IterateFromParams = {
  subIndex: string;
  fromId: number;
};

type CreatePersistentStorageOptions = {
  compression?: boolean;
  writable?: boolean;
};

/**
 * Persistent storage backed by SegmentedLog (RocksDB)
 * - Fast sequential appends
 * - Timestamp+sequence ordering
 * - Direct key overwrite support
 */
export function createPersistentStorage<T extends BaseStorageRecord>(
  baseDir: string,
  options?: CreatePersistentStorageOptions,
) {
  const dbCache = new Map<string, InstanceType<typeof SegmentedLog>>();
  const syncIntervals = new Map<string, NodeJS.Timeout>();

  const getDbPath = (subIndex: string): string => join(baseDir, subIndex);
  const getSecondaryPath = (subIndex: string): string => join(baseDir, `${subIndex}_secondary`);

  const getOrCreateDb = (subIndex: string): InstanceType<typeof SegmentedLog> => {
    const normalized = normalizeIndexDir(subIndex);
    const cached = dbCache.get(normalized);
    if (cached) return cached;

    const dbPath = getDbPath(normalized);
    mkdirSync(dirname(dbPath), { recursive: true });

    const db = options?.writable
      ? new SegmentedLog(dbPath, options?.compression ?? true)
      : SegmentedLog.openAsSecondary(
          dbPath,
          getSecondaryPath(normalized),
          options?.compression ?? true,
        );

    dbCache.set(normalized, db);

    if (!options?.writable) {
      const interval = setInterval(() => {
        try {
          db.tryCatchUpWithPrimary();
        } catch (error) {
          console.error(`Failed to sync secondary instance for ${normalized}:`, error);
        }
      }, 500);
      syncIntervals.set(normalized, interval);
    }

    return db;
  };

  const closeAllDatabases = (): void => {
    for (const interval of syncIntervals.values()) {
      clearInterval(interval);
    }
    syncIntervals.clear();

    for (const db of dbCache.values()) {
      try {
        db.close();
      } catch (error) {
        console.error('Failed to close rocksdb', error);
      }
    }
    dbCache.clear();
  };

  return {
    /**
     * Append one record (auto key: timestamp + seq)
     */
    async appendRecord(params: WriteRecordParams<T>): Promise<number> {
      const { record, subIndex } = params;
      record.timestamp = record.timestamp ?? Date.now();
      const db = getOrCreateDb(subIndex);
      const key = db.append(Buffer.from(JSON.stringify(record)));

      return parseKey(key);
    },

    /**
     * Append multiple records
     */
    async appendRecords(params: WriteRecordsParams<T>): Promise<number[]> {
      const { records, subIndex } = params;
      const db = getOrCreateDb(subIndex);
      const now = Date.now();
      const results = db.appendBatch(
        records.map((record) => {
          record.timestamp = record.timestamp ?? now;
          return Buffer.from(JSON.stringify(record));
        }),
      );

      return results.map((r) => parseKey(r));
    },

    /**
     * Replace or insert a specific record at a known key
     */
    async replaceRecord(params: ReplaceRecordParams<T>): Promise<void> {
      const { record, subIndex, id: key } = params;
      const db = getOrCreateDb(subIndex);
      record.timestamp = record.timestamp ?? Date.now();
      db.put(key, Buffer.from(JSON.stringify(record)));
    },

    /**
     * Replace the last record if it exists, otherwise append a new one.
     */
    async replaceOrInsertLastRecord(params: ReplaceLastRecordParams<T>) {
      const { record, subIndex } = params;
      const db = getOrCreateDb(subIndex);
      record.timestamp = record.timestamp ?? Date.now();

      // Try to read the most recent record
      const [last] = db.readLast(1);
      let key = last?.key ? parseKey(last.key) : undefined;
      if (key) {
        db.put(key, Buffer.from(JSON.stringify(record)));
      } else {
        // Append if database empty
        key = parseKey(db.append(Buffer.from(JSON.stringify(record))));
      }

      return key;
    },

    async readFullPage(params: ReadFullPageParams): Promise<StorageRecordReturn<T>[]> {
      const { subIndex } = params;
      const db = getOrCreateDb(normalizeIndexDir(subIndex));

      const results: StorageRecordReturn<T>[] = [];
      const iter = db.iterateFrom(0);

      try {
        while (iter.hasNext()) {
          const result = iter.next();
          if (result) {
            const key = parseKey(result.key);
            const value = JSON.parse(result.value.toString());
            results.push({ ...value, id: key });
          }
        }
      } finally {
        iter.close();
      }

      return results;
    },

    async iterateFrom(params: IterateFromParams): Promise<PersistentStorageIterator<T>> {
      const { subIndex, fromId } = params;
      const db = getOrCreateDb(normalizeIndexDir(subIndex));
      const iter = db.iterateFrom(fromId);

      return {
        hasNext() {
          return iter.hasNext();
        },
        close() {
          return iter.close();
        },
        next() {
          const item = iter.next();
          if (!item) {
            return null;
          }
          const value = JSON.parse(item.value.toString());
          value.id = parseKey(item.key);
          return value;
        },
      };
    },

    /**
     * Read forward from key (bigint)
     */
    async readRecordsRange(params: ReadRangeParams): Promise<StorageRecordReturn<T>[]> {
      const { subIndex, fromId, count } = params;
      const db = getOrCreateDb(subIndex);

      const results: StorageRecordReturn<T>[] = [];
      const iter = db.iterateFrom(fromId);
      const maxCount = count ?? 1000;

      try {
        let readCount = 0;
        while (iter.hasNext() && readCount < maxCount) {
          const result = iter.next();
          if (result) {
            const id = parseKey(result.key);
            const value = JSON.parse(result.value.toString());
            results.push({ id, ...value });
            readCount++;
          }
        }
      } finally {
        iter.close();
      }

      return results;
    },

    /**
     * Read the most recent N records
     */
    async readLastRecords(subIndex: string, count = 1): Promise<StorageRecordReturn<T>[]> {
      const db = getOrCreateDb(subIndex);
      const entries = db.readLast(count);
      const result: StorageRecordReturn<T>[] = [];
      for (const { key, value } of entries) {
        result.push({ ...JSON.parse(value.toString()), id: parseKey(key) });
      }
      return result;
    },

    /**
     * Read just the last record (if exists)
     */
    async readLastRecord(subIndex: string): Promise<StorageRecordReturn<T> | null> {
      const db = getOrCreateDb(subIndex);
      const [last] = db.readLast(1);
      if (!last) return null;
      return { ...JSON.parse(last.value.toString()), id: parseKey(last.key) };
    },

    close(): void {
      closeAllDatabases();
    },
  };
}

export type PersistentStorage<T extends BaseStorageRecord> = ReturnType<
  typeof createPersistentStorage<T>
>;

export function parseKey(key: Buffer): number {
  if (key.byteLength !== 8) {
    throw new Error(`Invalid key length ${key.byteLength}, expected 8 bytes`);
  }

  // Read big-endian 64-bit signed integer and convert to JS number
  const id = Number(key.readBigInt64BE(0));

  return id;
}

export type PersistentStorageIterator<T extends BaseStorageRecord> = {
  next(): StorageRecordReturn<T> | null;
  hasNext(): boolean;
  close(): void;
};

File: apps/py-predictor/src/lib/persistent_storage.py
import json
import re
from os import makedirs
from os.path import join
from typing import Protocol, TypedDict

from rocksdb_binding import SegmentedLog


class StorageRecord[T](TypedDict):
    timestamp: int
    id: int
    data: T


def normalize_sub_index(sub_index: str) -> str:
    normalized = sub_index.lower()
    normalized = re.sub(r"[^a-z0-9_]", "_", normalized)
    normalized = re.sub(r"_{2,}", "_", normalized)
    normalized = re.sub(r"^_+|_+$", "", normalized)
    return normalized


def serialize_key(id: int) -> bytes:
    return id.to_bytes(8, byteorder="big", signed=True)


def parse_key(key: bytes) -> int:
    return int.from_bytes(key, byteorder="big", signed=True)


class PersistentStorageIteratorRaw(Protocol):
    def next_batch(self):
        pass

    def has_next(self) -> bool:
        pass

    def close(self) -> None:
        pass


class PersistentStorageIterator[T]:
    def __init__(self, raw_iterator):
        self._raw_iterator = raw_iterator
        self._closed = False

    def next_batch(self) -> list[T]:
        if self._closed:
            return []

        raw_items = self._raw_iterator.next_batch()
        results: list[T] = []

        for item in raw_items:
            key_int = parse_key(bytes(item.key))
            data = json.loads(bytes(item.value).decode())
            data["id"] = key_int
            results.append(data)

        return results

    def has_next(self) -> bool:
        if self._closed:
            return False
        return self._raw_iterator.has_next()

    def close(self) -> None:
        if not self._closed:
            self._raw_iterator.close()
            self._closed = True


class PersistentStorage[T]:
    def __init__(self, base_dir: str, sub_index: str, writable: bool = True):
        self._base_dir = base_dir
        self._sub_index = normalize_sub_index(sub_index)
        self._writable = writable
        self._log: SegmentedLog | None = None

    def _primary_dir(self) -> str:
        return join(self._base_dir, self._sub_index)

    def _secondary_dir(self) -> str:
        return join(self._base_dir, self._sub_index + "_secondary")

    def _get_or_create_log(self) -> SegmentedLog:
        if self._log is not None:
            return self._log

        primary_dir = self._primary_dir()
        try:
            makedirs(primary_dir, exist_ok=True)
        except Exception:
            pass

        if self._writable:
            self._log = SegmentedLog(primary_dir, True)
        else:
            secondary_dir = self._secondary_dir()
            try:
                makedirs(secondary_dir, exist_ok=True)
            except Exception:
                pass
            self._log = SegmentedLog.open_as_secondary(primary_dir, secondary_dir, True)

        return self._log

    def append_record(self, record: T) -> int:
        log = self._get_or_create_log()
        payload = json.dumps(record).encode()
        key = log.append(payload)
        assigned_id = parse_key(bytes(key))
        try:
            record["id"] = assigned_id
        except Exception:
            pass
        return assigned_id

    def append_records(self, records: list[T]) -> list[int]:
        log = self._get_or_create_log()
        messages = [json.dumps(r).encode() for r in records]
        keys = log.append_batch(messages)
        ids: list[int] = []
        for i, k in enumerate(keys):
            assigned_id = parse_key(bytes(k))
            ids.append(assigned_id)
            try:
                records[i]["id"] = assigned_id
            except Exception:
                pass
        return ids

    def iterate_from(self, from_index: int, batch_size: int) -> PersistentStorageIterator[T]:
        log = self._get_or_create_log()
        raw_it = log.iterate_from(serialize_key(from_index), batch_size)
        return PersistentStorageIterator[T](raw_it)

    def iterate_from_raw(self, from_index: int, batch_size: int) -> PersistentStorageIteratorRaw:
        log = self._get_or_create_log()
        return log.iterate_from(serialize_key(from_index), batch_size)

    def read_records_range(self, from_index: int, count: int) -> list[T]:
        it = self.iterate_from(from_index, count)
        try:
            items = it.next_batch()
            return items[:count]
        finally:
            it.close()

    def read_last_record(self) -> T | None:
        log = self._get_or_create_log()
        items = log.read_last(1)
        if not items:
            return None
        item = items[0]
        key_int = parse_key(bytes(item.key))
        data = json.loads(bytes(item.value).decode())
        data["id"] = key_int
        return data

    def replace_or_insert_last_record(self, record: T) -> int:
        log = self._get_or_create_log()
        last_key = log.get_last_key()
        payload = json.dumps(record).encode()
        if last_key is None:
            key = log.append(payload)
            assigned_id = parse_key(bytes(key))
            try:
                record["id"] = assigned_id
            except Exception:
                pass
            return assigned_id
        last_id = parse_key(bytes(last_key))
        log.put(serialize_key(last_id), payload)
        try:
            record["id"] = last_id
        except Exception:
            pass
        return last_id

    def close(self) -> None:
        if self._log is not None:
            try:
                self._log.close()
            except Exception:
                pass
            self._log = None
